### PR TITLE
Issue 57

### LINKED ISSUE #57: Download update from custom url
Hi!
How can I upload my update not through Google Play service?
Thank you!

### PR DESCRIPTION
**Pull request by btkelly**
This should resolve #57 as it provides a way to download an update file as well as specify any custom action to take when a user selects update from the dialog util.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -155,6 +155,23 @@ new Gandalf.Installer()
         .install();
 ```
 
+#### Custom OnUpdateSelectedListener
+
+You my provide a custom listener to be invoked when the user selects to update their app. This can be helpful if you are not hosting your application on Google Play and would like to download an APK from another source. Two listeners are already provided, the `PlayStoreUpdateListener` which opens Google Play to the specified package name and the `FileDownloadUpdateListener` which will download a file specified by the Uri provided.
```
Comment (stkent) [inline README.md]: my -> may

Comment (stkent) [inline README.md]: Two listeners -> Two default listeners

File: README.md [hunk 2]
```diff
@@ -155,6 +155,23 @@ new Gandalf.Installer()
...
+
+```java
+new Gandalf.Installer()
+        .setContext(this)
+        .setOnUpdateSelectedListener(new OnUpdateSelectedListener() {
+            @Override
+            public void onSelectedUpdate(@NonNull Activity activity) {
```
Comment (stkent) [inline README.md]: onUpdateSelected would be closer to standard Android conventions

File: gandalf/src/main/java/io/github/btkelly/gandalf/utils/BootstrapDialogUtil.java [hunk 3]
```diff
@@ -41,13 +41,13 @@ private BootstrapDialogUtil() {
 
     public static void showRequiredUpdateDialog(@NonNull final Activity activity,
                                                 @NonNull final Gandalf gandalf,
-                                                @NonNull RequiredUpdate requiredUpdate) {
+                                                @NonNull final RequiredUpdate requiredUpdate) {
```
Comment (stkent) [inline gandalf/src/main/java/io/github/btkelly/gandalf/utils/BootstrapDialogUtil.java]: 🏆

File: gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    public void onSelectedUpdate(@NonNull Activity activity) {
+        downloadFile();
+        if (ActivityStateUtil.isActivityValid(activity)) {
+            activity.finish();
+        }
+    }
+
+    @Nullable
+    protected Map<String, String> requestHeaders() {
+        return null;
```
Comment (stkent) [inline gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java]: Needed?

File: gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    private void downloadFile() {
+        String fileName = fileUri.getLastPathSegment();
+
+        DownloadManager.Request request = new DownloadManager.Request(fileUri)
+                .setTitle(fileName)
+                .setMimeType(mimeType())
+                .setDestinationInExternalFilesDir(context, null, fileName)
+                .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
+
+        if (requestHeaders() != null) {
```
Comment (stkent) [inline gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java]: Needed?

**Comment by stkent** [comment]
@btkelly comments in!

File: README.md (hunk 2 above)
Comment (jsibbold) [inline README.md]: classic removeGlobalOnLayoutListener situation

File: gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java (hunk 5 above)
Comment (btkelly) [inline gandalf/src/main/java/io/github/btkelly/gandalf/utils/FileDownloadUpdateListener.java]: Yeah I would like to leave these in here if someone needs to add custom authorization headers. This was required in a previous DownloadManager use of mine so I wanted to make sure it's covered.

File: README.md (hunk 2 above)
Comment (btkelly) [inline README.md]: classic

Comment (stkent) [inline README.md]: legen- wait for it -dary

**Comment by stkent** [comment]
Need to update Travis config file for build tools version?

**Comment by btkelly** [comment]
Yup, totally forgot to do that. Good catch

**Comment by stkent** [comment]
Approved

**Pull request merged**